Add ability to invite circles to events

onny opened the pull request:
Fixes https://github.com/nextcloud/calendar/issues/2954

![signal-2023-08-04-161347_002](https://github.com/nextcloud/calendar/assets/757752/68bfc51d-62fe-46b8-8bd6-f571aea6657d)

Work done

- [ ] Frontend: Ability to search for circles
- [ ] Frontend: Ability to add circles to events
- [ ] Backend: Resolve circle members to individual participants
- [ ] Add tests
- [ ] Add release notes

ChristophWurst commented on src/components/Editor/Invitees/InviteesListSearch.vue:
```diff
@@ -237,6 +240,30 @@ export default {
+				return {
+					commonName: circle.displayname,
+					calendarUserType: circle.calendarUserType,
+					email: circle.email,
```
so conceptually it fetches the circle, expands to the individuals and adds everyone with an email address as attendee, right?

onny replied:
Have to figure it out too 😂

onny commented:
This is how the interface is for group invites on Google Calendar

![image](https://github.com/nextcloud/calendar/assets/757752/96589a92-1632-4792-9442-92920f57b195)

And this is the Google ical/ics file

```
BEGIN:VEVENT
DTSTART;VALUE=DATE:20230816
DTEND;VALUE=DATE:20230817
DTSTAMP:20230815T174303Z
ORGANIZER;CN=onnuex:mailto:onny@project-insanity.org
UID:37gsf88cc8****@google.com
ATTENDEE;CUTYPE=INDIVIDUAL;ROLE=REQ-PARTICIPANT;PARTSTAT=ACCEPTED;CN=onnuex
 ;X-NUM-GUESTS=0:mailto:onnyonnuex@gmail.com
ATTENDEE;CUTYPE=INDIVIDUAL;ROLE=REQ-PARTICIPANT;PARTSTAT=NEEDS-ACTION;CN=mygroup;X-NUM-GUESTS=0:mailto:mygroupgooglegroups.com
CREATED:20230815T173809Z
LAST-MODIFIED:20230815T173809Z
SEQUENCE:0
STATUS:CONFIRMED
SUMMARY:test
TRANSP:TRANSPARENT
END:VEVENT
```

onny commented:
Before further polishing this PR I would like to know what you think of this approach. It already works quite good but I think it can be optimized in some aspects

onny commented:
@miaulalala Maybe you can take a look at it if this could be a feasible approach, thank you :)

miaulalala commented:
cc @jancborchardt @nimishavijay 

I like it a lot, and the design to me looks really well thought through too.

tcitworld commented:
I like it too. I'm not sure « invitation sent » (the attendee status) makes sense on the group itself though.

I think we might need to tell the user some way or another that the member list might not be up-to-date with the upstream source (circles, NC groups, contact groups, …).

onny commented:
Thank you all for the feedback :) Looking forward to fix remaining issues mid September after my holidays.

@jancborchardt Would have really liked to join the conference but unfortunately I can't take part on this day :(

onny commented:
> Looks really nice and simple design-wise! The only suggestion I have is that instead of just "(2)" it could show "(2 members)" to be more clear, otherwise it looks like what is appended to a duplicate name or so. :)

Thanks this makes sense, fixed it

![image](https://github.com/nextcloud/calendar/assets/757752/ad165e22-3c11-43c1-88f7-c9760ba553bd)

> I like it too. I'm not sure « invitation sent » (the attendee status) makes sense on the group itself though.

fixed :)

> I also think we might need to tell the user some way or another that the member list might not be up-to-date with the upstream source (circles, NC groups, contact groups, …).

I thought about displaying a note after adding a group or circle to the attendee list (which would be done in a [separate PR](https://github.com/nextcloud/calendar/pull/5420))

![image](https://github.com/nextcloud/calendar/assets/757752/e3aea6df-921c-431c-90c3-d3eea08cf103)

onny commented:
From my side this PR is ready for a second review :) I added a detailed commit description on the first commit to explain all the changes:

```
Implement ability to render groups and group members as attendees of an
calendar event. This is implemented by reading and interpreting the
CUTYPE and MEMBER parameter of the iCal ATTENDEE attribute.

For example attendee "user1" belongs to a group identified with the uri
mygroup@localhost: "ATTENDEE;MEMBER="mailto:mygroup@localhost";CN=user1;[...]".
The group entity is also a ATTENDEE entry but with "CUTYPE=GROUP" and
with the corresponding uri mygroup@localhost:
"ATTENDEE;CN=MyGroup;CUTYPE=GROUP;[...]:mailto:mygroup@localhost".

In the frontend, groups will be displayed as a collapsible list whereas
members of the group will be displayed with the same functionality as
normal attendees.

In case the group entry gets deleted, all of its members will get
removed too.

Empty groups with no members might be present in the iCal file but won't
get rendered.

According to the iCal specifications, groups can have more than one uri and
attendees could have multiple membership addresses in one attribute.
This is neither covered yet in calendar-js nor in this PR.
```

onny commented:
This is the current state of both PRs combined (rendering groups & invite circles) 🥳 

![nextcloud-calendar-invite-circles](https://github.com/nextcloud/calendar/assets/757752/462ad595-a5dd-4f33-8761-90c389bb866f)

onny commented:
@miaulalala @jancborchardt @nimishavijay @tcitworld I tested and simplified the PR a bit more, would be nice to know if there's anything missing :)

miaulalala reviewed (approved):
Nice, I love it!

nimishavijay reviewed (approved):
Looks great design-wise :)

onny commented:
JS tests fail because I was using calendar-js master for support of the attendee.member attribute :)

st3iny commented:
@onny The calendar-js update is merged. Please rebase and we should be ready for the merge.

Ref https://github.com/nextcloud/calendar/pull/5562

onny commented:
thank you @st3iny. Oh the tests "codecov/patch" and "codecov/project" seem to fail. Do I have to look into it?

miaulalala commented:
That's fine, it's just the coverage. Let's get this merged!

miaulalala commented:
TYSM onny!
